scripts/download.sh

```diff
@@ -3,27 +3,79 @@
 set -eo pipefail
 
 CTLSTORE_BOOTSTRAP_URL=$1
+PREFIX="$(echo $CTLSTORE_BOOTSTRAP_URL | grep :// | sed -e's,^\(.*://\).*,\1,g')"
+URL="$(echo $CTLSTORE_BOOTSTRAP_URL | sed -e s,$PREFIX,,g)"
+BUCKET="$(echo $URL | grep / | cut -d/ -f1)"
+KEY="$(echo $URL | grep / | cut -d/ -f2)"
 CONCURRENCY=${2:-20}
 DOWNLOADED="false"
 COMPRESSED="false"
 METRICS="/var/spool/ctlstore/metrics.json"
 
 START=$(date +%s)
 END=$(date +%s)
+
+download_snapshot() {
+  s5cmd -r 0 --log debug cp --concurrency $CONCURRENCY $CTLSTORE_BOOTSTRAP_URL .
+}
+
+get_remote_checksum() {
+  remote_checksum=$(aws s3api head-object --bucket "${BUCKET}" --key "${KEY}" | jq -r '.Metadata.checksum // empty')
+  echo "$remote_checksum"
+}
+
 if [ ! -f /var/spool/ctlstore/ldb.db ]; then
   # busybox does not support sub-second resolution
   START=$(date +%s)
 
   mkdir -p /var/spool/ctlstore
   cd /var/spool/ctlstore
-  s5cmd -r 0 --log debug cp --concurrency $CONCURRENCY $CTLSTORE_BOOTSTRAP_URL .
 
-  DOWNLOADED="true"
-  if [[ ${CTLSTORE_BOOTSTRAP_URL: -2} == gz ]]; then
-    echo "Decompressing"
-    pigz -d snapshot.db.gz
-    COMPRESSED="true"
-  fi
+  COUNTER=0
+  while true; do
+    COUNTER=$(($COUNTER+1))
+
+    echo "Downloading head object from ${CTLSTORE_BOOTSTRAP_URL}"
+    checksum_before=$(get_remote_checksum)
+    echo "Remote checksum before downloading snapshot: $checksum_before"
+
+    echo "Downloading snapshot from ${CTLSTORE_BOOTSTRAP_URL}"
+    download_snapshot
+
+    echo "Downloading head object from ${CTLSTORE_BOOTSTRAP_URL}"
+    checksum_after=$(get_remote_checksum)
+    echo "Remote checksum after downloading snapshot: $checksum_after"
```

Comment:
The reason why I am trying to download the header object twice is sort of a workaround to a potential atomicity issue. Due to `s5cmd`'s lack of support of downloading the header object which contains the checksum value, we have to issue a separate s3 api call to retrieve it, and ctlstore-supervisor uploads a new snapshot every 5 minutes, there is chances that the creation of new snapshot happens just between the downloading of header object and the snapshot, it would return false negative.

Reply:
If we have object versioning turned on, `s5cmd` supports a `version-id` parameter to download a specific version of the object. If we make a `head-object` call and fetch both the SHA and the object version (which is in the `x-amz-version-id` header), we can fetch the exact version of the object we received the checksum for and avoid the atomicity issue.

Reply:
Interesting, @erikdw also mentioned about object versioning but he was afraid "it could be expensive". Reading [Using versioning in S3 buckets](https://docs.aws.amazon.com/AmazonS3/latest/userguide/Versioning.html?icmpid=docs_amazons3_console) a bit and did some experiments in a testing bucket I could see that it's feasible to create a lifecycle rule to delete older version of objects so cost should not be a concern:
<img width="813" alt="Screenshot 2023-09-29 at 10 34 42 AM" src="https://github.com/segmentio/ctlstore/assets/6007569/d80c60fe-98b2-42b0-99fc-55f9207808b5">

Reply:
I can see another benefits of doing so is to be an alternative solution to a question Taylor once asked: https://twilio.slack.com/archives/C0324FBR4Q5/p1691769186456639. But yea I think this is totally a more elegant way to handle this! I'll wait for a +1 from Erik before iterate on it.

Reply:
I like the idea in general, it's more inline with what I originally assumed such a verification would need to look like before I learned about the idea you've currently implemented.